src/stash.rs

```diff
@@ -0,0 +1,204 @@
+            opts.configure(&mut self.raw_opts.checkout_options);
+        }
+        self.checkout_options = Some(opts);
+        self
+    }
+
+    /// Optional callback to notify the consumer of application progress.
+    ///
+    /// Return `true` to continue processing, or `false` to
+    /// abort the stash application.
+    pub fn progress_cb<C>(&mut self, callback: C) -> &mut StashApplyOptions<'cb>
+        where C: FnMut(StashApplyProgress) -> bool + 'cb
+    {
+        self.progress = Some(Box::new(callback) as Box<StashApplyProgressCb<'cb>>);
+        self.raw_opts.progress_cb = stash_apply_progress_cb;
+        self.raw_opts.progress_payload = self as *mut _ as *mut _;
+        self
+    }
+
+    /// Pointer to a raw git_stash_apply_options
+    pub fn raw(&self) -> &raw::git_stash_apply_options {
+        &self.raw_opts
+    }
+}
+
+#[allow(unused)]
+pub struct StashCbData<'a> {
+    pub callback: Box<StashCb<'a>>
+}
+
+#[allow(unused)]
+pub extern fn stash_cb(index: size_t,
+                   message: *const c_char,
+                   stash_id: *const raw::git_oid,
+                   payload: *mut c_void)
+                   -> c_int
+{
+    panic::wrap(|| unsafe {
+        let mut data = &mut *(payload as *mut StashCbData);
+        let res = {
+            let mut callback = &mut data.callback;
+            callback(
+                index,
+                CStr::from_ptr(message).to_str().unwrap(),
+                &Binding::from_raw(stash_id)
+            )
+        };
+
+        if res { 0 } else { 1 }
+    }).unwrap()
+}
+
+fn convert_progress(progress: raw::git_stash_apply_progress_t) -> StashApplyProgress {
+    match progress {
+        raw::GIT_STASH_APPLY_PROGRESS_NONE => StashApplyProgress::None,
+        raw::GIT_STASH_APPLY_PROGRESS_LOADING_STASH => StashApplyProgress::LoadingStash,
+        raw::GIT_STASH_APPLY_PROGRESS_ANALYZE_INDEX => StashApplyProgress::AnalyzeIndex,
+        raw::GIT_STASH_APPLY_PROGRESS_ANALYZE_MODIFIED => StashApplyProgress::AnalyzeModified,
+        raw::GIT_STASH_APPLY_PROGRESS_ANALYZE_UNTRACKED => StashApplyProgress::AnalyzeUntracked,
+        raw::GIT_STASH_APPLY_PROGRESS_CHECKOUT_UNTRACKED => StashApplyProgress::CheckoutUntracked,
+        raw::GIT_STASH_APPLY_PROGRESS_CHECKOUT_MODIFIED => StashApplyProgress::CheckoutModified,
+        raw::GIT_STASH_APPLY_PROGRESS_DONE => StashApplyProgress::Done,
+
+        _ => StashApplyProgress::None
+    }
+}
+
+#[allow(unused)]
+extern fn stash_apply_progress_cb(progress: raw::git_stash_apply_progress_t,
+                                  payload: *mut c_void)
+                                  -> c_int
+{
+    panic::wrap(|| unsafe {
+        let mut options = &mut *(payload as *mut StashApplyOptions);
+        let res = {
+            let mut callback = options.progress.as_mut().unwrap();
+            callback(convert_progress(progress))
+        };
+
+        if res { 0 } else { -1 }
+    }).unwrap()
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
@alexcrichton should I move tests to `repo.rs` as well? Or it's better to leave them here?